### PR TITLE
epoll

### PR DESCRIPTION
**Pull request by tanner0101**
- [x] `TranslatingStream` helper for creating AsymmetricStreams
- [x] TransformingStream -> `TranscribingStream`
- [x] `EpollEventLoop`
- [x] `DefaultEventLoop` type that points to appropriate event loop
- [x] kqueue and epoll cleaned up to use c void* ptr baton for holding event source reference

### CODE REVIEW

File: Sources/Async/EventLoop/EventLoop.swift [hunk 1]
```diff
@@ -28,6 +36,10 @@ public protocol EventLoop: Worker {
     /// is ready to write data and the event source is resumed.
     func onWritable(descriptor: Int32, _ callback: @escaping EventCallback) -> EventSource
 
+    /// Creates a new timer event source.
+    /// This callback will be called perodically when the timeout is reached.
+    func onTimeout(timeout: Int, _ callback: @escaping EventLoop.EventCallback) -> EventSource
```
Comment (Joannis) [inline Sources/Async/EventLoop/EventLoop.swift]: What is the `timout: Int`? Is it seconds, MS or Nanoseconds even?

File: Sources/Async/EventLoop/Epoll/EpollEventSource.swift [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+            let tfd = timerfd_create(CLOCK_MONOTONIC, 0)
+            if tfd == -1 {
+                fatalError("timerfd_create() failed: errno=\(errno)")
+            }
+
+            var ts = itimerspec()
+            ts.it_interval.tv_sec = 0;
+            ts.it_interval.tv_nsec = 0;
+            ts.it_value.tv_sec = timeout / 1000;
+            ts.it_value.tv_nsec = (timeout % 1000) * 1000000;
```
Comment (Joannis) [inline Sources/Async/EventLoop/Epoll/EpollEventSource.swift]: semicolons from copied C code? 🎉

File: Sources/Async/EventLoop/EventLoop.swift [hunk 3]
```diff
@@ -59,6 +71,18 @@ extension EventLoop {
     }
 }
 
+extension Thread {
+    public static func async(_ work: @escaping () -> ()) {
+        if #available(OSX 10.12, *) {
+            Thread.detachNewThread {
+                work()
```
Comment (Joannis) [inline Sources/Async/EventLoop/EventLoop.swift:78]: Related: https://github.com/vapor/async/pull/27

File: Sources/Async/EventLoop/Kqueue/KqueueEventLoop.swift [hunk 4]
```diff
@@ -41,89 +32,80 @@ public final class KqueueEventLoop: EventLoop {
...
-
         // check for new events
-        let eventCount = kevent(kq, nil, 0, eventlist.baseAddress, Int32(eventlist.count), &timeout)
+        let eventCount = kevent(kq, nil, 0, eventlist.baseAddress, Int32(eventlist.count), nil)
         guard eventCount >= 0 else {
-            print("An error occured while running kevent: \(eventCount).")
-            return
+            switch errno {
+            case EINTR:
+                run() // run again
```
Comment (Joannis) [inline Sources/Async/EventLoop/Kqueue/KqueueEventLoop.swift:85]: Is this really the best way to handle interrupts?

File: Sources/Async/EventLoop/Kqueue/KqueueEventLoop.swift [hunk 5]
```diff
@@ -41,89 +32,80 @@ public final class KqueueEventLoop: EventLoop {
...
+                return
+            default:
+                let reason = String(cString: strerror(Int32(errno)))
+                print("An error occured while running kevent: \(reason).")
```
Comment (Joannis) [inline Sources/Async/EventLoop/Kqueue/KqueueEventLoop.swift]: This triggers on breakpoints, too

File: Sources/Async/EventLoop/Kqueue/KqueueEventSource.swift [hunk 6]
```diff
@@ -92,6 +122,10 @@ public final class KqueueEventSource: EventSource {
             }
         }
     }
+
+    deinit {
+        print("\(type(of: self)).\(#function)")
```
Comment (Joannis) [inline Sources/Async/EventLoop/Kqueue/KqueueEventSource.swift]: print?

File: Sources/Async/EventLoop/Kqueue/KqueueEventLoop.swift (hunk 4 above)
Comment (tanner0101) [reply Sources/Async/EventLoop/Kqueue/KqueueEventLoop.swift:85]: > When kevent() call	fails with EINTR error,	all changes in the changelist
     have been applied.

From this I believe so.

File: Sources/Async/EventLoop/EventLoop.swift (hunk 1 above)
Comment (tanner0101) [reply Sources/Async/EventLoop/EventLoop.swift]: milliseconds. I will change the naming here to be more clear!

**Pull request merged**